cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,495 @@
+  int indices_per_source,
+  const rmm::device_uvector<typename GraphViewType::edge_type>& global_degree_offsets)
+{
+  using vertex_t  = typename GraphViewType::vertex_type;
+  using edge_t    = typename GraphViewType::edge_type;
+  auto edge_count = active_sources_in_row.size() * indices_per_source;
+  rmm::device_uvector<vertex_t> sources(edge_count, handle.get_stream());
+  rmm::device_uvector<vertex_t> destinations(edge_count, handle.get_stream());
+  rmm::device_uvector<gpu_t> destination_gpu_ids(edge_count, handle.get_stream());
+
+  auto [vertex_id_seg, vertex_count_offsets, graph_offsets, graph_indices] =
+    detail::partition_information(handle, graph_view);
+
+  thrust::for_each(
+    handle.get_thrust_policy(),
+    thrust::make_counting_iterator<size_t>(0),
+    thrust::make_counting_iterator<size_t>(edge_count),
+    [edge_index_first,
+     active_sources        = active_sources_in_row.data(),
+     active_source_gpu_ids = active_source_gpu_ids.data(),
+     vertex_id_seg         = vertex_id_seg.data(),
+     vertex_id_seg_count   = vertex_id_seg.size(),
+     vertex_count_offsets  = vertex_count_offsets.data(),
+     global_degree_offsets = global_degree_offsets.data(),
+     graph_offsets         = graph_offsets.data(),
+     graph_indices         = graph_indices.data(),
+     sources               = sources.data(),
+     destinations          = destinations.data(),
+     dst_gpu_ids           = destination_gpu_ids.data(),
+     invalid_vertex_id,
+     indices_per_source] __device__(auto index) {
+      // source which this edge index refers to
+      auto loc           = index / indices_per_source;
+      auto source        = active_sources[loc];
+      sources[index]     = source;
+      dst_gpu_ids[index] = active_source_gpu_ids[loc];
+
+      // Find which partition id did the source belong to
+      auto partition_id =
+        thrust::distance(
+          vertex_id_seg,
+          thrust::upper_bound(
+            thrust::seq, vertex_id_seg, vertex_id_seg + vertex_id_seg_count, source)) -
+        1;
+      // starting position of the segment within global_degree_offset
+      // where the information for partition (partition_id) starts
+      //  vertex_count_offsets[partition_id]
+      // The relative location of offset information for vertex id v within
+      // the segment
+      //  v - seg[partition_id]
+      auto location_in_segment = source - vertex_id_seg[partition_id];
+
+      // csr offset value for vertex v that belongs to partition (partition_id)
+      auto csr_offset       = graph_offsets[partition_id][location_in_segment];
+      auto local_out_degree = graph_offsets[partition_id][location_in_segment + 1] - csr_offset;
+      edge_t const* adjacency_list = graph_indices[partition_id] + csr_offset;
+      // read location of global_degree_offset needs to take into account the
+      // partition offsets because it is a concatenation of all the offsets
+      // across all partitions
```

Comment:
And should we really iterate over each edge?

This is beneficial comparing to iterating over vertices if vertex degrees vary a lot, but here loop count per vertex is `indices_per_source` so no thread-divergence issue.